File: day04/part1.lua
local function in_bounds(grid, x, y)
  return x >= 1 and x <= #grid and y >= 1 and y <= #grid[1]
end

local function count(grid)
  local directions = {
    { dx = 1,  dy = 0 },  -- Right
    { dx = -1, dy = 0 },  -- Left
    { dx = 0,  dy = 1 },  -- Down
    { dx = 0,  dy = -1 }, -- Up
    { dx = 1,  dy = 1 },  -- Diagonal down-right
    { dx = -1, dy = -1 }, -- Diagonal up-left
    { dx = 1,  dy = -1 }, -- Diagonal down-left
    { dx = -1, dy = 1 },  -- Diagonal up-right
  }

  local word = "XMAS"
  local word_length = #word
  local word_count = 0

  for x = 1, #grid do
    for y = 1, #grid[1] do
      for _, dir in ipairs(directions) do
        local matches = true
        for i = 0, word_length - 1 do
          local nx, ny = x + i * dir.dx, y + i * dir.dy
          if not in_bounds(grid, nx, ny) or grid[nx]:sub(ny, ny) ~= word:sub(i + 1, i + 1) then
            matches = false
            break
          end
        end
        if matches then
          word_count = word_count + 1
        end
      end
    end
  end

  return word_count
end

local function solve(filename)
  local grid = {}
  for line in io.lines(filename) do
    table.insert(grid, line)
  end
  return count(grid)
end

print(solve("input.txt"))

File: day04/part1.exs
defmodule Part1 do
  @word "XMAS"
  @directions [
    {1, 0},   # Right
    {-1, 0},  # Left
    {0, 1},   # Down
    {0, -1},  # Up
    {1, 1},   # Diagonal down-right
    {-1, -1}, # Diagonal up-left
    {1, -1},  # Diagonal down-left
    {-1, 1}   # Diagonal up-right
  ]

  defp count_matches(grid) do
    Enum.reduce(0..(length(grid) - 1), 0, fn x, acc ->
      Enum.reduce(0..(String.length(Enum.at(grid, x)) - 1), acc, fn y, acc2 ->
        acc2 + count_word(grid, x, y)
      end)
    end)
  end

  defp count_word(grid, x, y) do
    Enum.reduce(@directions, 0, fn {dx, dy}, acc ->
      if word_matches?(grid, x, y, dx, dy), do: acc + 1, else: acc
    end)
  end

  defp word_matches?(grid, x, y, dx, dy) do
    @word
    |> String.graphemes()
    |> Enum.with_index()
    |> Enum.all?(fn {char, i} ->
      nx = x + i * dx
      ny = y + i * dy
      in_bounds?(grid, nx, ny) and String.at(Enum.at(grid, nx), ny) == char
    end)
  end

  defp in_bounds?(grid, x, y) do
    x >= 0 and x < length(grid) and y >= 0 and y < String.length(Enum.at(grid, x))
  end

  def solve(filename) do
    filename
    |> File.stream!()
    |> Enum.map(&String.trim/1)
    |> count_matches()
  end
end

"input.txt"
|> Part1.solve()
|> IO.puts()
